File: projects/MusicGame/Music-Game-main/src/Player.java
import java.awt.*;
import javax.swing.Timer; // Import for Timer
import java.awt.event.ActionListener; // Import for ActionListener
import java.awt.event.ActionEvent; // Import for ActionEvent

// The Player class represents the player in the game.
public class Player extends Control {
    // Constants for the player's size, gravity, jump strength, and speed
    public static final int SIZE = 21;
    private static final double GRAVITY = 0.35;
    static final double JUMP_STRENGTH = 7.5;
    private static final int SPEED = 4;

    // Variables for the player's fall speed for after powerUp collision
    private double fallSpeed = GRAVITY;

    //PowerUp constants
    private static final int BOOSTED_SPEED = 7;
    private boolean canCollideWithPaddle = true; // boolean to check if the player can collide with the paddle for the PowerUp logic

    // Variables for the player's vertical velocity, score, direction, and color
    private double velocityY;
    private int score = 0;
    private int direction = 0;

    // Constructor for the Player class
    public Player(int x, int y, int velocityX, int velocityY) {
        super(x, y, velocityX, velocityY, SIZE, SIZE);
        this.velocityY = velocityY;
    }

    // Method to increase the player's score
    public void increaseScore() {
        score++;
    }

    // Method to apply gravity to the player
    public void applyGravity() {
        velocityY += GRAVITY;
    }

    public void slowFall() {
        fallSpeed = GRAVITY - .1; // Slow fall speed

        // Create a timer to reset the fall speed after 5 seconds
        Timer resetFallSpeedTimer = new Timer(2000, new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                fallSpeed = GRAVITY; // Reset to default fall speed
            }
        });
        resetFallSpeedTimer.setRepeats(false);
        resetFallSpeedTimer.start();
    }

    // Method to reverse the player's vertical velocity to simulate a bounce
    public void bounce() {
        velocityY = -JUMP_STRENGTH;
    }

    // Method to move the player based on velocity and direction
    public void move() {
        setY((int) (getY() + velocityY));
        setX(getX() + direction * SPEED);
    }

    // Method to move the player to the left
    public void moveLeft() {
        direction = -1;
    }

    // Method to move the player to the right
    public void moveRight() {
        direction = 1;
    }

    // Method to stop the player's horizontal movement
    public void stopMoving() {
        direction = 0;
    }

    // Method to draw the player
    @Override
    public void draw(Graphics g) {
        g.setColor(new Color(0x664321));
        g.fillRect(x, y, SIZE, SIZE);
    }

    // Method to check if the player is colliding with a paddle
    public boolean isCollidingWithPaddle(Paddle paddle) {
        boolean isColliding = getBounds().intersects(paddle.getBounds());
        if (isColliding && canCollideWithPaddle) {
            bounce();
        }
        return isColliding;
    }

    // Method to check if the player has fallen off the screen
    public boolean isOutOfBounds(int gameHeight) {
        return getY() > gameHeight;
    }
    // Method to check if player went off the left side
    public boolean wentoffscreenLeft(){
        return getX() < 0;
    }
    // Method to check if player went off the right side
    public boolean wentoffscreenRight(int gameWidth){
        return getX() > gameWidth;
    }

    // Method to handle a collision with a platform
    public void handleCollision() {
        increaseScore();
    }

    // Method to get the player's score
    public int getScore() {
        return score;
    }

    // Method to get the player's bounding rectangle
    public Rectangle getBounds() {
        return new Rectangle(getX(), getY(), SIZE, SIZE);
    }

    // Getter and setter methods for the player's x and y coordinates
    public int getX() {
        return x;
    }
    
    public int getY() {
        return y;
    }
    
    public void setX(int newX) {
        this.x = newX;
    }
    
    public void setY(int newY) {
        this.y = newY;
    }
    
    public void addScore(int score) {
        this.score += score;
    }

    //PowerUp methods
    public void speedUp() {
        this.velocityX = BOOSTED_SPEED;
        this.canCollideWithPaddle = false;
    }
    
    public void resetSpeed() {
        this.velocityX = SPEED;
        this.canCollideWithPaddle = true;
    }
    
    public void stopVerticalMovement() {
        this.velocityY = 0;
    }
    
    public void resumeVerticalMovement() {
        this.velocityY = SPEED;
    }

    public int setScore(int i) {
        return this.score;
    }
}

File: projects/MusicGame/Music-Game-main/src/Player.js
class Player extends Control {
    static SIZE = 21;
    static GRAVITY = 0.35;
    static JUMP_STRENGTH = 7.5;
    static SPEED = 4;

    constructor(x, y, velocityX, velocityY) {
        super(x, y, velocityX, velocityY, Player.SIZE, Player.SIZE);
        this.velocityY = velocityY;
        this.fallSpeed = Player.GRAVITY;
        this.canCollideWithPaddle = true;
        this.score = 0;
        this.direction = 0;
    }

    increaseScore() {
        this.score++;
    }

    applyGravity() {
        this.velocityY += Player.GRAVITY;
    }

    bounce() {
        this.velocityY = -Player.JUMP_STRENGTH;
    }

    move() {
        this.applyGravity();
        this.setY(this.getY() + this.velocityY);
        this.setX(this.getX() + this.direction * Player.SPEED);
    }

    moveLeft() {
        this.direction = -1;
    }

    moveRight() {
        this.direction = 1;
    }

    stopMoving() {
        this.direction = 0;
    }

    draw(ctx) {
        ctx.fillStyle = '#664321';
        ctx.fillRect(this.x, this.y, Player.SIZE, Player.SIZE);
    }

    isCollidingWithPaddle(paddle) {
        const isColliding = this.getBounds().intersects(paddle.getBounds());
        if (isColliding && this.canCollideWithPaddle) {
            this.canCollideWithPaddle = false;
            setTimeout(() => this.canCollideWithPaddle = true, 500);
        }
        return isColliding;
    }

    getBounds() {
        return {
            x: this.getX(),
            y: this.getY(),
            width: Player.SIZE,
            height: Player.SIZE,
            intersects: function (rect) {
                return (
                    this.x < rect.x + rect.width &&
                    this.x + this.width > rect.x &&
                    this.y < rect.y + rect.height &&
                    this.y + this.height > rect.y
                );
            }
        };
    }

    getX() {
        return this.x;
    }

    getY() {
        return this.y;
    }

    setX(newX) {
        this.x = newX;
    }

    setY(newY) {
        this.y = newY;
    }
}

class Paddle extends Control {
    static HEIGHT = 20;
    static WIDTH = 70;
    static NORMAL_SPEED = 2;

    constructor(x, y) {
        super(x, y, 0, 0, Paddle.WIDTH, Paddle.HEIGHT);
        this.dy = Paddle.NORMAL_SPEED;
    }

    draw(ctx) {
        ctx.fillStyle = '#D2B48C';
        ctx.fillRect(this.x, this.y, Paddle.WIDTH, Paddle.HEIGHT);
    }

    moveDown(bottomBound, gameWidth) {
        this.y += this.dy;
        if (this.y > bottomBound) {
            this.y = 0;
            this.x = Math.floor(Math.random() * (gameWidth - Paddle.WIDTH));
        }
    }

    getBounds() {
        return {
            x: this.x,
            y: this.y,
            width: Paddle.WIDTH,
            height: Paddle.HEIGHT,
            intersects: function (rect) {
                return (
                    this.x < rect.x + rect.width &&
                    this.x + this.width > rect.x &&
                    this.y < rect.y + rect.height &&
                    this.y + this.height > rect.y
                );
            }
        };
    }
}